pandas/core/arrays/floating.py

```diff
@@ -0,0 +1,641 @@
+    _FloatingDtype. For example we have Float32Dtype to represent float32.
+
+    The attributes name & type are set when these subclasses are created.
+    """
+
+    name: str
+    base = None
+    type: Type
+
+    def __repr__(self) -> str:
+        return f"{self.name}Dtype()"
+
+    @property
+    def _is_numeric(self) -> bool:
+        return True
+
+    @cache_readonly
+    def numpy_dtype(self) -> np.dtype:
+        """ Return an instance of our numpy dtype """
+        return np.dtype(self.type)
+
+    @cache_readonly
+    def kind(self) -> str:
+        return self.numpy_dtype.kind
+
+    @cache_readonly
+    def itemsize(self) -> int:
+        """ Return the number of bytes in this dtype """
+        return self.numpy_dtype.itemsize
+
+    @classmethod
+    def construct_array_type(cls) -> Type["FloatingArray"]:
+        """
+        Return the array type associated with this dtype.
+
+        Returns
+        -------
+        type
+        """
+        return FloatingArray
+
+    def _get_common_dtype(self, dtypes: List[DtypeObj]) -> Optional[DtypeObj]:
+        # for now only handle other floating types
+        if not all(isinstance(t, _FloatingDtype) for t in dtypes):
+            return None
+        np_dtype = np.find_common_type(
+            [t.numpy_dtype for t in dtypes], []  # type: ignore
+        )
+        if np.issubdtype(np_dtype, np.floating):
+            return _dtypes[str(np_dtype)]
+        return None
+
+    def __from_arrow__(
+        self, array: Union["pyarrow.Array", "pyarrow.ChunkedArray"]
+    ) -> "FloatingArray":
+        """
+        Construct FloatingArray from pyarrow Array/ChunkedArray.
+        """
+        import pyarrow  # noqa: F811
+        from pandas.core.arrays._arrow_utils import pyarrow_array_to_numpy_and_mask
+
+        pyarrow_type = pyarrow.from_numpy_dtype(self.type)
+        if not array.type.equals(pyarrow_type):
+            array = array.cast(pyarrow_type)
+
+        if isinstance(array, pyarrow.Array):
+            chunks = [array]
+        else:
+            # pyarrow.ChunkedArray
+            chunks = array.chunks
+
+        results = []
+        for arr in chunks:
+            data, mask = pyarrow_array_to_numpy_and_mask(arr, dtype=self.type)
+            float_arr = FloatingArray(data.copy(), ~mask, copy=False)
+            results.append(float_arr)
+
+        return FloatingArray._concat_same_type(results)
+
+
+def coerce_to_array(
+    values, dtype=None, mask=None, copy: bool = False,
+) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    Coerce the input values array to numpy arrays with a mask
+
+    Parameters
+    ----------
+    values : 1D list-like
+    dtype : float dtype
+    mask : bool 1D array, optional
+    copy : bool, default False
+        if True, copy the input
+
+    Returns
+    -------
+    tuple of (values, mask)
+    """
+    # if values is floating numpy array, preserve it's dtype
+    if dtype is None and hasattr(values, "dtype"):
```

Comment:
Is it possible to share this code with the integer implementation? Wondering if we shouldn't have a numeric base class instead of copy / pasting

Reply:
to @WillAyd point; it makes sense to just add things that are already not in the super class (or put them there if possible) notw that we have multiple arrays classes that can share; it might be worth creating a NumericArray that is the parent of Integer and Floating (and inherits from BaseMaskedArray)

Reply:
And I already answered extensively on this here: https://github.com/pandas-dev/pandas/pull/34307#issuecomment-636829484